File: video_10_minPathSumInGrid.cpp
// #include<iostream>
// #include<bits/stdc++.h>
// using namespace std;
/**************************************************** MEMOIZATION ******************************************************** */

// int solve(int row, int col, vector<vector<int>> &grid, vector<vector<int>> &dp){
//     if(row == 0 && col == 0) return grid[row][col];
//     if(row<0 || col<0) return 1e9;
//     if(dp[row][col] != -1) return dp[row][col];

//     int up = grid[row][col]+solve(row-1, col, grid, dp);
//     int left = grid[row][col]+solve(row, col-1, grid, dp);

//     return dp[row][col] = min(up, left);
// }
// int minSumPath(vector<vector<int>> &grid) {
//     // Write your code here.
//     int row = grid.size();
//     int col = grid[0].size();
//     vector<vector<int>> dp(row, vector<int> (col, -1));
//     //dp[0][0] = 1;
//     return solve(row-1, col-1, grid, dp);
// }
/*************************************************** TABULATION ********************************************************* */
// int minSumPath(vector<vector<int>> &matrix) {
//     // Write your code here.
//     int n = matrix.size();
//     int m = matrix[0].size();
//     vector<vector<int>> dp(n, vector<int>(m, 0)); // Initialize a DP table to store minimum path sums

//     for (int i = 0; i < n; i++) {
//         for (int j = 0; j < m; j++) {
//             if (i == 0 && j == 0)
//                 dp[i][j] = matrix[i][j]; 
//             else {
//                 int up = matrix[i][j];
//                 if (i > 0) up += dp[i - 1][j]; 
//                 else up += 1e9; 

//                 int left = matrix[i][j];
//                 if (j > 0) left += dp[i][j - 1]; 
//                 else left += 1e9; 

//                 dp[i][j] = min(up, left);
//             }
//         }
//     }return dp[n-1][m-1];
// }

/************************************************* MAIN *************************************************************** */
// int main(){
//     int row, col, elements;
//     cout<<"enter rows and col: ";
//     cin>>row>>col;
//     vector<vector<int>> grid(row, vector<int> (col));
//     cout<<"enter elements in the grid: ";
//     for(int i = 0; i<row; i++){
//         for(int j = 0; j<col; j++){
//             cin>>grid[i][j];
//         }
//     }
//     cout<<"min path sum is: "<<minSumPath(grid);
//     return 0;
// }

File: video_10_minPathSumInGrid.java
import java.util.*;
class video_10_minPathSumInGrid{
    //Function to return the minimum cost to react at bottom
	//right cell from top left cell.
    public static int[][] dp, vis;
    public static int ans;

    public static void solve(int sr, int sc, int row ,int col, int[][] grid, int pathLen){
        if(sr==row && sc==col){
            ans = Math.min(ans, pathLen+grid[row][col]);
            return;
        }
        
        if(dp[sr][sc] != -1) {
            pathLen += dp[sr][sc];
            return;
        }
        
        vis[sr][sc] = 1;
        int drow[] = {1, -1, 0, 0};
        int dcol[] = {0, 0, 1, -1};

        for(int i = 0; i<4; i++){
            int nrow = sr+drow[i];
            int ncol = sc+dcol[i];

            if(nrow>=0 && nrow<=row && ncol>=0 && ncol<=col){
                if(vis[nrow][ncol] != 1)
                solve(nrow, ncol, row, col, grid, grid[row][col]+pathLen);
            }
        }
        vis[sr][sc] = 0;
        dp[sr][sc] = pathLen;
    }
    
    //Function to return the minimum cost to react at bottom
	//right cell from top left cell.
    public static int minimumCostPath(int[][] grid){
        // Code here
        int row = grid.length-1;
        int col = grid[0].length-1;
        ans = Integer.MAX_VALUE;

        vis = new int[row+1][col+1];
        for(int i = 0; i<=row; i++){
            for(int j = 0; j<=col; j++)
                vis[i][j] = 0;
        }

        dp = new int[row+1][col+1];
        for(int i = 0; i<=row; i++){
            for(int j = 0; j<=col; j++)
                dp[i][j] = -1;
        }

        solve(0, 0, row, col, grid, 0);
        return ans;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int row = sc.nextInt();
        int col = sc.nextInt();

        int grid[][] = new int[row][col];
        for(int i = 0; i<row; i++){
            for(int j = 0; j<col; j++)
                grid[i][j] = sc.nextInt();
        }

        System.out.print(minimumCostPath(grid));
        System.out.println();
        //System.out.print(dp[row-1][col-1]);
        for(int i = 0; i<row; i++){
            for(int j = 0; j<col; j++)
                System.out.print(dp[i][j]+" ");
            System.out.println();
        }
    }
}